File: fonter/cachy/src/services/ProjectService.cpp
#include "ProjectService.h"

#include "../core/EditorTypes.h"

#include <QFile>
#include <QHash>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>

ProjectService::ProjectService(QObject *parent) : QObject(parent) {}

bool ProjectService::saveProject(const QString &path, const EditorStore &store, QString *errorMessage) const {
    if (path.trimmed().isEmpty()) {
        if (errorMessage) *errorMessage = QStringLiteral("Project path is empty.");
        return false;
    }
    QJsonObject root;
    root.insert("formatVersion", 2);
    root.insert("strokePx", store.strokePx());
    root.insert("captureGapUm", store.captureGapUm());
    root.insert("zoom", store.zoom());
    root.insert("eraseRadiusPx", store.eraseRadiusPx());

    QJsonArray strokes;
    for (const Stroke &stroke : store.strokes()) {
        QJsonObject s;
        s.insert("id", stroke.id);
        s.insert("createdAt", QString::number(stroke.createdAt));
        QJsonArray points;
        for (const Stroke::StrokePoint &pt : stroke.points) {
            QJsonObject p;
            p.insert("x", pt.pos.x());
            p.insert("y", pt.pos.y());
            p.insert("erased", pt.erased);
            points.push_back(p);
        }
        s.insert("points", points);
        strokes.push_back(s);
    }
    root.insert("strokes", strokes);

    QJsonArray boxes;
    for (const SelectionBox &box : store.selectionBoxes()) {
        QJsonObject b;
        b.insert("id", box.id);
        b.insert("orderIndex", box.orderIndex);
        b.insert("x", box.rect.x);
        b.insert("y", box.rect.y);
        b.insert("width", box.rect.width);
        b.insert("height", box.rect.height);
        b.insert("assigned", box.assigned);
        b.insert("assignedAscii", box.assignedAscii);
        b.insert("fileStem", box.fileStem);
        b.insert("joinMode", joinModeToString(box.joinMode));
        b.insert("hasManualAnchor", box.hasManualAnchor);
        b.insert("manualAnchorRx", box.manualAnchorRx);
        b.insert("manualAnchorRy", box.manualAnchorRy);
        b.insert("anchorX", box.anchorX);
        b.insert("anchorY", box.anchorY);
        boxes.push_back(b);
    }
    root.insert("selectionBoxes", boxes);
    root.insert("selectedSelectionId", store.selectedSelectionId());
    QJsonObject selectionErasedPoints;
    for (auto it = store.selectionErasedPointKeys().cbegin(); it != store.selectionErasedPointKeys().cend(); ++it) {
        QJsonArray keys;
        for (const QString &key : it.value()) keys.push_back(key);
        selectionErasedPoints.insert(it.key(), keys);
    }
    root.insert("selectionErasedPoints", selectionErasedPoints);

    QJsonObject stems;
    for (auto it = store.specialCharStemMap().cbegin(); it != store.specialCharStemMap().cend(); ++it) {
        stems.insert(QString::number(it.key()), it.value());
    }
    root.insert("specialCharStemMap", stems);

    QFile file(path);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Truncate | QIODevice::Text)) {
        if (errorMessage) *errorMessage = file.errorString();
        return false;
    }
    file.write(QJsonDocument(root).toJson(QJsonDocument::Indented));
    file.close();
    return true;
}

bool ProjectService::loadProject(const QString &path, EditorStore *store, QString *errorMessage) const {
    if (!store) {
        if (errorMessage) *errorMessage = QStringLiteral("Store is null.");
        return false;
    }
    QFile file(path);
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        if (errorMessage) *errorMessage = file.errorString();
        return false;
    }
    const QJsonDocument doc = QJsonDocument::fromJson(file.readAll());
    file.close();
    if (!doc.isObject()) {
        if (errorMessage) *errorMessage = QStringLiteral("Invalid project file.");
        return false;
    }
    const QJsonObject root = doc.object();
    const int fmt = root.value("formatVersion").toInt(0);
    if (fmt != 1 && fmt != 2) {
        if (errorMessage) *errorMessage = QStringLiteral("Unsupported project version.");
        return false;
    }

    store->clearAll();
    store->setStrokePx(root.value("strokePx").toInt(store->strokePx()));
    store->setCaptureGapUm(root.value("captureGapUm").toInt(store->captureGapUm()));
    store->setZoom(root.value("zoom").toInt(store->zoom()));
    store->setEraseRadiusPx(root.value("eraseRadiusPx").toInt(store->eraseRadiusPx()));

    // Recreate strokes
    QVector<Stroke> loadedStrokes;
    const QJsonArray strokes = root.value("strokes").toArray();
    for (const QJsonValue &sv : strokes) {
        const QJsonObject s = sv.toObject();
        const QJsonArray points = s.value("points").toArray();
        if (points.isEmpty()) continue;
        Stroke stroke;
        stroke.id = s.value("id").toString();
        stroke.createdAt = s.value("createdAt").toString().toLongLong();
        stroke.points.reserve(points.size());
        for (const QJsonValue &pv : points) {
            const QJsonObject p = pv.toObject();
            Stroke::StrokePoint point;
            point.pos = QPointF(p.value("x").toDouble(), p.value("y").toDouble());
            point.erased = p.value("erased").toBool(false);
            stroke.points.push_back(point);
        }
        loadedStrokes.push_back(stroke);
    }
    store->setStrokes(loadedStrokes);

    // Recreate boxes
    QVector<SelectionBox> loadedBoxes;
    const QJsonArray boxes = root.value("selectionBoxes").toArray();
    for (const QJsonValue &bv : boxes) {
        const QJsonObject b = bv.toObject();
        SelectionBox box;
        box.id = b.value("id").toString();
        box.orderIndex = b.value("orderIndex").toInt(0);
        box.rect = SelectionRect{
            b.value("x").toDouble(),
            b.value("y").toDouble(),
            b.value("width").toDouble(),
            b.value("height").toDouble()};
        box.assigned = b.value("assigned").toBool(false);
        box.assignedAscii = b.value("assignedAscii").toInt(-1);
        box.fileStem = b.value("fileStem").toString();
        box.joinMode = joinModeFromString(b.value("joinMode").toString());
        box.hasManualAnchor = b.value("hasManualAnchor").toBool(false);
        box.manualAnchorRx = b.value("manualAnchorRx").toDouble(0.0);
        box.manualAnchorRy = b.value("manualAnchorRy").toDouble(0.0);
        box.anchorX = b.value("anchorX").toDouble(0.0);
        box.anchorY = b.value("anchorY").toDouble(0.0);
        loadedBoxes.push_back(box);
    }
    store->setSelectionBoxes(loadedBoxes, root.value("selectedSelectionId").toString());
    QHash<QString, QSet<QString>> selectionErasedPoints;
    const QJsonObject selectionErasedObj = root.value("selectionErasedPoints").toObject();
    for (auto it = selectionErasedObj.begin(); it != selectionErasedObj.end(); ++it) {
        const QJsonArray keys = it.value().toArray();
        QSet<QString> keySet;
        for (const QJsonValue &kv : keys) {
            const QString key = kv.toString();
            if (!key.isEmpty()) keySet.insert(key);
        }
        if (!keySet.isEmpty()) selectionErasedPoints.insert(it.key(), keySet);
    }
    store->setSelectionErasedPointKeys(selectionErasedPoints);

    QHash<int, QString> stemMap;
    const QJsonObject stemObj = root.value("specialCharStemMap").toObject();
    for (auto it = stemObj.begin(); it != stemObj.end(); ++it) stemMap.insert(it.key().toInt(), it.value().toString());
    store->setSpecialCharStemMap(stemMap);
    store->setProjectFilePath(path);
    store->markSaved();
    return true;
}

File: fonter/web/src/logic/projectService.ts
/**
 * Mirrors fonter/cachy/src/services/ProjectService.cpp
 * (file I/O replaced by string / object in/out for the web).
 */

import type { EditorStore } from "./editorStore";
import {
  joinModeFromString,
  joinModeToString,
  type SelectionBox,
  type Stroke,
  type StrokePoint,
} from "./editorTypes";

export interface HwProjectJson {
  formatVersion: number;
  strokePx: number;
  captureGapUm: number;
  guideLineGapPx?: number;
  zoom: number;
  eraseRadiusPx: number;
  strokes: Array<{
    id: string;
    createdAt: string;
    points: Array<{ x: number; y: number; erased?: boolean }>;
  }>;
  selectionBoxes: Array<{
    id: string;
    orderIndex: number;
    x: number;
    y: number;
    width: number;
    height: number;
    assigned: boolean;
    assignedAscii: number;
    fileStem: string;
    joinMode: string;
    hasManualAnchor?: boolean;
    manualAnchorRx?: number;
    manualAnchorRy?: number;
    anchorX: number;
    anchorY: number;
  }>;
  selectedSelectionId: string;
  /** selectionId -> strokeId#pointIndex keys */
  selectionErasedPoints?: Record<string, string[]>;
  specialCharStemMap: Record<string, string>;
}

export function saveProjectToJson(store: EditorStore): string {
  const erased = store.getSelectionErasedPointKeys();
  const selectionErasedPoints: Record<string, string[]> = {};
  for (const [selId, keys] of erased) {
    selectionErasedPoints[selId] = [...keys];
  }

  const root: HwProjectJson = {
    formatVersion: 2,
    strokePx: store.strokePx(),
    captureGapUm: store.captureGapUm(),
    guideLineGapPx: store.guideLineGapPx(),
    zoom: store.zoom(),
    eraseRadiusPx: store.eraseRadiusPx(),
    strokes: [],
    selectionBoxes: [],
    selectedSelectionId: store.selectedSelectionId(),
    selectionErasedPoints,
    specialCharStemMap: {},
  };

  for (const stroke of store.strokes()) {
    const points = stroke.points.map((pt) => ({
      x: pt.pos.x,
      y: pt.pos.y,
      erased: pt.erased,
    }));
    root.strokes.push({
      id: stroke.id,
      createdAt: String(stroke.createdAt),
      points,
    });
  }

  for (const box of store.selectionBoxes()) {
    root.selectionBoxes.push({
      id: box.id,
      orderIndex: box.orderIndex,
      x: box.rect.x,
      y: box.rect.y,
      width: box.rect.width,
      height: box.rect.height,
      assigned: box.assigned,
      assignedAscii: box.assignedAscii,
      fileStem: box.fileStem,
      joinMode: joinModeToString(box.joinMode),
      hasManualAnchor: box.hasManualAnchor,
      manualAnchorRx: box.manualAnchorRx,
      manualAnchorRy: box.manualAnchorRy,
      anchorX: box.anchorX,
      anchorY: box.anchorY,
    });
  }

  const stems = store.getSpecialCharStemMap();
  for (const [k, v] of stems) {
    root.specialCharStemMap[String(k)] = v;
  }

  return JSON.stringify(root, null, 2);
}

export function loadProjectFromJson(
  store: EditorStore,
  jsonText: string,
  projectPathForName: string,
): { ok: true } | { ok: false; error: string } {
  let root: HwProjectJson;
  try {
    root = JSON.parse(jsonText) as HwProjectJson;
  } catch {
    return { ok: false, error: "Invalid project file." };
  }

  if (!root || typeof root !== "object") {
    return { ok: false, error: "Invalid project file." };
  }

  const fmt = root.formatVersion ?? 0;
  if (fmt !== 1 && fmt !== 2) {
    return { ok: false, error: "Unsupported project version." };
  }

  store.clearAll();
  store.setStrokePx(root.strokePx ?? store.strokePx());
  store.setCaptureGapUm(root.captureGapUm ?? store.captureGapUm());
  store.setGuideLineGapPx(root.guideLineGapPx ?? store.guideLineGapPx());
  store.setZoom(root.zoom ?? store.zoom());
  store.setEraseRadiusPx(root.eraseRadiusPx ?? store.eraseRadiusPx());

  const loadedStrokes: Stroke[] = [];
  for (const s of root.strokes ?? []) {
    const pointsArr = s.points ?? [];
    if (pointsArr.length === 0) continue;
    const points: StrokePoint[] = [];
    for (const p of pointsArr) {
      points.push({
        pos: { x: p.x, y: p.y },
        erased: p.erased ?? false,
      });
    }
    loadedStrokes.push({
      id: s.id,
      createdAt: Number(s.createdAt),
      points,
    });
  }
  store.setStrokes(loadedStrokes);

  const loadedBoxes: SelectionBox[] = [];
  for (const b of root.selectionBoxes ?? []) {
    loadedBoxes.push({
      id: b.id,
      orderIndex: b.orderIndex ?? 0,
      rect: {
        x: b.x,
        y: b.y,
        width: b.width,
        height: b.height,
      },
      assigned: b.assigned ?? false,
      assignedAscii: b.assignedAscii ?? -1,
      fileStem: b.fileStem ?? "",
      joinMode: joinModeFromString(b.joinMode ?? "N"),
      hasManualAnchor: b.hasManualAnchor ?? false,
      manualAnchorRx: b.manualAnchorRx ?? 0,
      manualAnchorRy: b.manualAnchorRy ?? 0,
      anchorX: b.anchorX ?? 0,
      anchorY: b.anchorY ?? 0,
    });
  }
  store.setSelectionBoxes(loadedBoxes, root.selectedSelectionId ?? "");

  const erasedObj = root.selectionErasedPoints ?? {};
  const erasedMap = new Map<string, Set<string>>();
  for (const key of Object.keys(erasedObj)) {
    const arr = erasedObj[key];
    if (Array.isArray(arr) && arr.length > 0) {
      erasedMap.set(key, new Set(arr.filter((x) => typeof x === "string")));
    }
  }
  store.setSelectionErasedPointKeys(erasedMap);

  const stemMap = new Map<number, string>();
  const stemObj = root.specialCharStemMap ?? {};
  for (const key of Object.keys(stemObj)) {
    stemMap.set(Number(key), stemObj[key]!);
  }
  store.setSpecialCharStemMap(stemMap);

  store.setProjectFilePath(projectPathForName);
  store.markSaved();
  return { ok: true };
}
